proposed/clock-meta.md

```diff
@@ -0,0 +1,154 @@
+# PHPDoc Meta Document
+
+## 1. Summary
+
+The purpose of using the ClockInterface is to provide a standard way to access the system 
+time, that would allow interopability when testing code that relies on the current time 
+rather than relying installing PHP extensions or use hacks like re-declaring the `time()`
+function in other namespaces. 
+
+## 2. Why Bother?
+
+There are currently a few libraries that do provide the functionality on packagist, however 
+there is no interopability between these different libraries, as they ship with their own 
+clock interfaces. Symfony has a TimeMock library which uses namespace hacks to override the 
+`time()`, `date()`, `microtime()`, etc functions, however this does not solve mocking calls to 
+`new \DateTime()`
+
+Pros:
+
+* Consistent interface to get the current time
+* Easy to mock the wall clock time for repeatablility.
+
+Cons:
+
+* Extra overhead and developer effort to get the current time, not as simple as
+calling `time()` or `date()`.
+
+## 3. Scope
+
+### 3.1 Goals
+
+* Provide a simple and mockable way to read the current time
+* Allow interoperability between libraries when reading the clock
+
+### 3.2 Non-Goals
+
+* This PSR does not provide a recommendation on how and when to use the concepts
+  described in this document, so it is not a coding standard.
+* This PSR does not provide a recommendation on how to handle timezones when 
+  retrieving the current time. This is left up to the implementation.
+
+## 4. Approaches
+
+### 4.1 Chosen Approach
+
+We have decided to formalize the existing practices, use by several other packages
+out in the wild. Some of the popular packages providing this functionality are: 
+lcobucci/clock, kreait/clock, ergebnis/clock, and mangoweb/clock. Some providing
+interfaces, and some relying on overloading (extending) the Clock class to mock the
+current time.
+
+
+### 4.2 Example Implemntations
+
+```php
+final class TimeZoneAwareClock implements \Psr\Clock\ClockInterface
```

Comment:
I definitely feel very uncomfortable with this `TimeZoneAwareClock` as the first implementation shown as an example.

It feels like the normal case to implement it would be to provide a timezone, while this is a very bad practice to promote. time-zone-based code leads to a code you can't scale easily worldwide, you can't easily move consistently to an other place, not easily re-use.

We should show first the `UTCClock` then simply show that timezone can be changed on the returned object:

```php
$universalNow = (new UTCClock())->now();
$nowForMyChicagoUser = $universalNow->setTimezone(new DateTimeZone('America/Chicago'));
```

And we should encourage to use the first one to do any back-end stuff or calculation, while reserving the second kind only for the very output to be displayed to an end user.

Reply:
@kylekatarnls

Why not inject the timezone via constructor?

```php
final class SystemClock implements Clock
{
    private \DateTimeZone $timezone;

    public function __construct(\DateTimeZone $timezone)
    {
        $this->timezone = $timezone;
    }

    public function now(): \DateTimeImmutable
    {
        return new \DateTimeImmutable(
            'now',
            $this->timezone
        );
    }

    public function freeze(): FrozenClock
    {
        return new FrozenClock($this->now());
    }
}
```

See https://github.com/ergebnis/clock/blob/2.2.0/src/SystemClock.php#L16-L37.

Reply:
I don't see the point this is exactly the same as the current proposal for `TimeZoneAwareClock` and showing this to people is the same as promoting "You can use any timezone for your clock". AND. THAT'S. NOT. GOOD.

As the person who might have to call `->now()` on an object implementing `ClockInterface` I want to get something reliable, not something in unknown, dynamic or even just some hard coded city timezone name.

If someone want a timezone they can do it outside of the clock using `->setTimezone()` and as this PSR says it's not aimed to say how to deal with timezone, then it should not show clocks that handle it internally.

Sure users still will be able to do it. But we should encourage them to do it outside the clock. So you give the clock only the responsibility to tick the moment, then give to some other service the responsibility to tell you what numbers are given to this tick in a given timezone.

Reply:
You get something very reliable: A `DateTimeImmutable` Object that represents the current time!

What you will never be able to tell is, whether the current time was observed in Cairo, Sydney, Lima or London. That is something that your own business logic needs to handle anyhow.

The whole `ClockInterface` will be completely timezone agnostic.

Reply:
Throwing my opinion in here since we’re still going back and forth on timezones.

+1 to encouraging UTC and leaving timezone manipulation to app implementation.  IMHO if someone needs this they can create their own interface, the PSR should enforce best practice whenever possible.

Reply:
This PSR is timezone-agnostic.

Therefore we should **not** encourage anything in regards to UTC or any other timezone. 

BestPractices in timezone-handling is to use `DateTimeImmutable` which we are enforcing. The internals of that `DateTieImmutable`-Instance are not our business.

Reply:
Perfect, thanks @heiglandreas